File: src/main/scala/no/mesan/fag/reactive/scala/actors/SideAnsvarlig.scala
package no.mesan.fag.reactive.scala.actors

import akka.actor.{ActorLogging, Props, ActorRef, Actor}

/** Holder orden på sidene vi har scannet, starter nye scans ved behov. */
class SideAnsvarlig(bildeAnsvarlig: ActorRef, domain: String) extends Actor with ActorLogging with Tracer {
  traceCreate(s"SideAnsvarlig for $bildeAnsvarlig i $domain")

  def withFoundPages(foundPages: Set[String]): Receive = {
    case scan: ScanMessage =>
      def isANewPage(url: String) = !(foundPages contains url)
      def isInSameDomain(url: String) = url contains domain
      val url = scan.url.replaceAll("[?].*$", "")  // Ta bort queryparam

      if (isANewPage(url) && isInSameDomain(url)) {
        scan.traceReceive(this)
        val reader = context.actorOf(SideLeser.props(bildeAnsvarlig))
        reader ! scan
      }
      else scan.traceReceive("ignoring")

      context become withFoundPages(foundPages + url)
  }

  override def receive = withFoundPages(Set.empty)
}

/** Opprettelse og meldinger. */
object SideAnsvarlig {
  def props(bildeAnsvarlig: ActorRef, domain: String) =
    Props(classOf[SideAnsvarlig], bildeAnsvarlig, domain)
}

File: src/main/java/no/mesan/reactive/actors/SideAnsvarlig.java
package no.mesan.reactive.actors;

import java.util.HashSet;
import java.util.Set;

import no.mesan.reactive.actors.messages.ScanMessage;

import akka.actor.AbstractActor;
import akka.actor.ActorRef;
import akka.actor.Props;
import akka.japi.pf.ReceiveBuilder;

public class SideAnsvarlig extends AbstractActor {

    private final String domain;
    private Set<String> foundPages = new HashSet<>();
    private ActorRef bildeAnsvarlig;

    public SideAnsvarlig(final ActorRef bildeAnsvarlig, final String domain) {
        this.bildeAnsvarlig = bildeAnsvarlig;
        this.domain = domain;

        receive(ReceiveBuilder
          .match(ScanMessage.class, this::handleScanMessage)
          .matchAny((Object message) -> {
              unhandled(message);
          }).build()
        );
    }

    private void handleScanMessage(final ScanMessage scanMessage) {
        final String url = scanMessage.getUrl();
        if (notVisitedYet(url) && isInSameDomain(url)) {
            final ActorRef sideLeser = context().actorOf(Props.create(SideLeser.class, bildeAnsvarlig));
            sideLeser.tell(scanMessage, context().self());
        }
        foundPages.add(url);
    }

    private boolean notVisitedYet(final String url) {
        return !foundPages.contains(url);
    }

    private boolean isInSameDomain(final String url) {
        return url.contains(domain);
    }
}
